packages/models-library/src/models_library/rest_pagination_utils.py

```diff
@@ -0,0 +1,88 @@
+from math import ceil
+from typing import Any, Dict, List, Protocol, TypedDict, Union, runtime_checkable
+
+from .rest_pagination import PageLinks, PageMetaInfoLimitOffset
+
+# NOTE: In this repo we use two type of URL-like data structures:
+#  - from yarl (aiohttp-style) and
+#  - from starlette (fastapi-style)
+#
+# Here define protocol to avoid including starlette  or yarl in this librarie's requirements
+# and a helper function below that can handle both protocols at runtime
+#
+
+
+@runtime_checkable
+class _YarlURL(Protocol):
+    def update_query(self, query) -> "_YarlURL":
+        ...
+
+
+class _StarletteURL(Protocol):
+    # SEE starlette.data_structures.URL
+    #  in https://github.com/encode/starlette/blob/master/starlette/datastructures.py#L130
+
+    def replace_query_params(self, **kwargs: Any) -> "_StarletteURL":
```

Comment:
just wondering, are you replacing the fastapi-pagination with this? [https://github.com/uriyyo/fastapi-pagination](https://github.com/uriyyo/fastapi-pagination)
if yes, why?

Reply:
i was not aware we were using fastapi-pagination :-)

What i did here was adapting the pagination we have in place so it could be used with whatever URL-type we like. Yes, that would indeed allow using it to paginate with fastapi-like services.

That said, I do not think we should have different pagination structures for fastapi or aiohttp-like services. I rather keep this for any kind of service

Reply:
ahh i understand now your comment. That library uses exactly the same model name ``Page[*]`` ... :-) well, this was pure chance. In any case, it is quite obvious name and class layout (coming from C++ template programming). The rest was adapting the class you design to fit the generic. This PR basically adds a small wrapper around the design you did.